---
title: "assignment week 7"
author: "Timo Suter"
date: "2023-04-27"
output: html_document
---

What is wrong about the following R markdown?
```{r}
suppressPackageStartupMessages({
  library(epiwraps)
  library(AnnotationHub)
  library(MotifDb)
  library(universalmotif)
  library(ensembldb)
  library(ggplot2)
})
ah <- AnnotationHub (localHub = TRUE)
```

# download data
```{r}
options(timeout=3600)
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam", "atac.chr19.bam", mode = "wb")
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam.bai", "atac.chr19.bam.bai", mode = "wb")
```

# prepare tracks
```{r}
bam <- "atac.chr19.bam"
# create a track using all fragments, computing the coverage at each position (and force Ensembl naming)
bam2bw(bam, output_bw = "full_cov.bw", paired=TRUE, binWidth=10L, forceSeqlevelsStyle = "Ensembl")
# create a track using all fragments, but computing the number of cuts/insertion sites at each position
bam2bw(bam, output_bw = "full_cuts.bw", paired=TRUE, binWidth=1L, type="ends", forceSeqlevelsStyle = "Ensembl")
# create a track using only nucleosome-free fragments, computing coverage
bam2bw(bam, output_bw = "NF_cov.bw", paired=TRUE, binWidth=5L, minFragLength=30, 
       maxFragLength=120, forceSeqlevelsStyle = "Ensembl")
# create a track using only nucleosome-free fragments, the number of cuts/insertion sites at each position
bam2bw(bam, output_bw = "NF_cuts.bw", paired=TRUE, binWidth=1L, type="ends", minFragLength=30, 
       maxFragLength=120, forceSeqlevelsStyle = "Ensembl")
# create a track using only mono-nucleosome fragments, computing coverage
bam2bw(bam, output_bw = "mono_cov.bw", paired=TRUE, binWidth=10L, minFragLength=140,
       maxFragLength=220, forceSeqlevelsStyle = "Ensembl")
# create a track using only mono-nucleosome fragments, computing centers
bam2bw(bam, output_bw = "mono_centers.bw", paired=TRUE, binWidth=5L, minFragLength=140,
       maxFragLength=220, type="center", forceSeqlevelsStyle = "Ensembl")
```

# factor KLF4
## get genome & sequence for chr19
```{r}
genome <- ah[["AH68356"]]
chr19 <- import(genome)["19"]
```

## get KLF4 motif
```{r}
motif1_KLF4 <- query(MotifDb, c("KLF4","Mus"))[[1]]
motif2_KLF4 <- convert_motifs(motif1_KLF4, class = "TFBSTools-PFMatrix")
```

## find motif occurences
```{r}
moi_KLF4 <- motifmatchr::matchMotifs(motif2_KLF4, chr19, out = "positions", p.cutoff = 1e-5)[[1]]
moi_KLF4 <- as(setNames(moi_KLF4, names(chr19)), "GRanges")
```

## plotting signal around KLF4 motif occurences
```{r}
tracks <- c("Full coverage"="full_cov.bw", "NF coverage"="NF_cov.bw", "All cuts"="full_cuts.bw",
            "NF cuts"="NF_cuts.bw", "Mono cov"="mono_cov.bw", "Mono centers"="mono_centers.bw")
sm_KLF4 <- signal2Matrix(tracks, moi_KLF4, w = 5, extend = 500)
nf <- bwNormFactors(tracks, useSeqLevels = "19", nwind = 5000L)
smb_KLF4 <- rescaleSignalMatrices(sm_KLF4, nf)
plotEnrichedHeatmaps(smb_KLF4, trim = 1, minRowVal = 15, colors = c("white", "darkred"))
```

# factor MAZ
## get MAZ motif
```{r}
motif1_MAZ <- query(MotifDb, c("MAZ","Mus"))[[1]]
motif2_MAZ <- convert_motifs(motif1_MAZ, class = "TFBSTools-PFMatrix")
```

## find motif occurences
```{r}
moi_MAZ <- motifmatchr::matchMotifs(motif2_MAZ, chr19, out = "positions", p.cutoff = 1e-5)[[1]]
moi_MAZ <- as(setNames(moi_MAZ, names(chr19)), "GRanges")
```

## plotting signal around MAZ motif occurences
```{r}
sm_MAZ <- signal2Matrix(tracks, moi_MAZ, w = 5, extend = 500)
smb_MAZ <- rescaleSignalMatrices(sm_MAZ, nf)
plotEnrichedHeatmaps(smb_MAZ[1:5], trim = 0.95, minRowVal = 15, colors = c("white", "darkred"))
```
