---
title: "Working with DNA motifs"
author: "Pierre-Luc"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(GenomicRanges)
  library(ggplot2)
  library(memes) # for the meme-based methods -- COMMENT OUT when using alternatives
  library(motifmatchr)
  library(Biostrings) # for handling sequences
  library(MotifDb) # database of motifs
  library(TFBSTools) # for handling some motif formats
  library(universalmotif) # for converting motifs to various formats
  library(PWMEnrich) # for R-based motif enrichment analysis
})
```


# Obtaining peak sequences

For the purpose of this example, we'll use the CTCF peaks found on chromosome 1 of mESC:

Download:

```{r}
download.file("https://www.encodeproject.org/files/ENCFF508CKL/@@download/ENCFF508CKL.bed.gz", "mESC_CTCF_ENCFF508CKL.bed.gz")
peaks <- rtracklayer::import("mESC_CTCF_ENCFF508CKL.bed.gz", format="NarrowPeak")
seqlevelsStyle(peaks) <- "Ensembl"  # to change the convention of the chromosome names to ensembl (i.e. without 'chr')
peaks_chr1 <- peaks[seqnames(peaks)=="1"]
```



Since the motif of a transcription factor is typically found towards the center or summit of the peak, it is common practice to look for motifs around the center of peaks. 
How far around depends on the resolution of your data, i.e. how narrow are your peaks (can be between 50bp to 500bp around), as well as the nature of the signal from which the peaks were obtained.
The first thing we do, then, is to extract the regions around the peaks, and then obtain the sequence from those regions:

```{r}
peak_centers <- resize(peaks_chr1, fix="center", width=100)
# we get the genome sequence:
ah <- AnnotationHub(localHub=TRUE)
genome <- ah[["AH68356"]]

# we get the sequences corresponding to the peak centers:
peak_seqs <- memes::get_sequence(peak_centers, genome)
# peak_seqs <- Biostrings::getSeq(genome, peak_centers)
# names(peak_seqs) <- as.character(granges(peak_centers))
peak_seqs
as.character(peak_seqs[1])
```

# Motif scanning

Motif **scanning** aims at finding the _occurrences_ of a known motif in a set of sequences.
As this is methodologically fairly simple, it doesn't really matter what method one uses, though results may differ because of how the thresholds are set.

## Getting the desired motif

```{r}
# we search for "CTCF" in the motif database
motifs <- query(MotifDb, "CTCF")
# there are several matching motifs:
names(motifs)
# we select one:
motif <- motifs[["Hsapiens-JASPAR_CORE-CTCF-MA0139.1"]]
# we visualize it:
view_motifs(motif)
```


## Scanning a set of sequences using the Meme suite

```{r}
moi <- memes::runFimo(peak_seqs,  convert_motifs(motif))
# you might eventually need to add the meme_path argument:
moi <- memes::runFimo(peak_seqs,  convert_motifs(motif), meme_path = "/home/plger/bin/")
head(moi)
```


## Alternative using motifmatchr (doesn't require Meme installation):

```{r}
# if you don't already have the genome in fasta format saved somewhere, convert it to that format:
Biostrings::writeXStringSet(rtracklayer::import(genome), "genome.fa")
# we also need to convert the motif to a format that this package will accept
motif2 <- convert_motifs(motif, class="TFBSTools-PFMatrix")
# if we had a list of motifs instead, we'd convert them like this:
# motifs2 <- setNames(do.call(PFMatrixList, convert_motifs(motifs, class="TFBSTools-PFMatrix")), names(motifs))
moi2 <- motifmatchr::matchMotifs(motif2, subject=peak_centers, genome=Rsamtools::FaFile("genome.fa"), out="positions")[[1]]
head(moi2)
```

### Another R alternative...

For those users who can't install memes & where `motifmatchr::matchMotifs` somehow doesn't work...

```{r}
#' findMotifInstances
#'
#' A wrapper around TFBSTools for scanning motif occurence, handling the 
#' coordinate conversion as `memes` does.
#' 
#' @param seqs A set of sequences, e.g. `DNAStringSet`, optionally with 
#'   coordinate as names to enable conversion. (e.g. use 
#'   `names(seqs) <- as.character(granges(peaks))` to set them)
#' @param motif A motif, in any format recognized by `universalmotif`
#' @param keepMatchedSeq Logical; whether to keep the matched sequence.
#' @param ... Passed to `TFBSTools::searchSeq`; can for instance be used to set
#'   the number of threads to use, e.g. with `mc.cores=2`
#'
#' @return A `GRanges` object
findMotifInstances <- function(seqs, motif, keepMatchedSeq=FALSE, ...){
  motif <- universalmotif::convert_motifs(motif, "TFBSTools-PWMatrix")
  x <- suppressWarnings(TFBSTools::searchSeq(motif, subject=seqs))
  peaks <- strsplit(gsub("-",":",names(seqs)), ":")
  if(!is.null(names(seqs)) && all(lengths(peaks)==3)){ # convert relative coordinates to absolute
    chrs <- sapply(peaks,FUN=function(x) x[1])
    offsets <- sapply(peaks,FUN=function(x) as.integer(x[2]))-1L
    i <- rep(seq_along(x),lengths(x))
    return(GRanges(chrs[i], 
                   IRanges(offsets[i]+as.integer(unlist(lapply(x, start))),
                           offsets[i]+as.integer(unlist(lapply(x, end)))),
                   strand=unlist(lapply(x, strand)),
                   score=as.numeric(unlist(lapply(x, FUN=function(x) score(x))))
                   ))
  }
  x <- as(x, "GRanges")
  score(x) <- x$absScore
  keepFields <- c("score","relScore")
  if(keepMatchedSeq) keepFields <- c(keepFields, "siteSeqs")
  mcols(x) <- mcols(x)[,intersect(colnames(mcols(x)), keepFields)]
  x
}
```

```{r}
moi <- findMotifInstances(peak_seqs, motif)
```


## Comparison of the results

```{r}
epiwraps::regionUpset(list(fimo=moi, motifmatchr=moi2))
boxplot(moi$score~overlapsAny(moi, moi2))
```


```{r}
# peaks with motif:
peaks_w_motif <- peaks_chr1[overlapsAny(peaks_chr1,moi)]
length(peaks_w_motif)
length(peaks_chr1)
length(peaks_w_motif)/length(peaks_chr1)
```


## Scanning the whole genome

To scan the whole genome, one needs to use the genome sequence as input subject.
For example, using `motifmatchr::matchMotif` :

```{r}
# import the genome sequence:
genome_seqs <- import(genome)
# if your genome was a fasta file instead, import using Biostrings::readDNAStringSet()

# run the scan:
motif2 <- convert_motifs(motif, class="TFBSTools-PFMatrix")
motif_across_genome <- matchMotifs(motif2, subject=genome_seqs, out="positions")[[1]]
# when inputing matchMotifs directly with sequences, one has to do the following 
# to transform the output to GRanges:
names(motif_across_genome) <- names(genome_seqs)
motif_across_genome <- as(motif_across_genome, "GRanges")
head(motif_across_genome)
```




# Motif enrichment analysis

Motif **enrichment analysis** aims at finding _known_ motifs that are _enriched_ in a set of sequences (e.g. peaks) versus a background.
By default, the background is generated by shuffling the input sequences, but it can also be specified.



## Example using AME from the Meme suite

```{r}
# obtain the motifs (e.g. here use all the mouse motifs)
# To avoid having redundant motifs, we query motifs from a single source:
motifs <- query(MotifDb, c("Mmusculus", "HOCOMOCOv10"))
ame <- memes::runAme(peak_seqs, database=convert_motifs(motifs), meme_path="/home/plger/bin/")
head(ame)
```

We could also plot the results:

```{r}
ggplot(ame, aes(log2((1+tp)/(1+fp)), -log10(adj.pvalue), size=tp_percent)) + 
  geom_point(alpha=0.3) + geom_text(aes(label=motif_id), data=head(ame)) +
  labs(x="log2(fold-enrichment)", size="% of set")
```


## Example using PWMEnrich (much slower, but doesn't require memes)

```{r, eval=FALSE}
# prepare multithreading:
registerCoresPWMEnrich(4)
# launch the analysis:
res <- motifEnrichment(peak_seqs, convert_motifs(motifs, "PWMEnrich-PWM"))
# for the sake of example, we'll just use the first 300 sequences:
res <- motifEnrichment(head(peak_seqs, convert_motifs(motifs, "PWMEnrich-PWM"))
groupReport(res)
```





# Motif discovery

Motif **discovery** aims at finding _new motifs_ that are enriched in a set of sequences (e.g. peaks) versus a background (e.g. random or control sequences). This is the step where different methods will differ the most.

I'm giving you example code here, but since this is a computationally-intensive analysis we won't run it live.

## Example using Meme

```{r, eval=FALSE}
# this requires having the meme suite installed
mr <- memes::runMeme( peak_seqs,
                      p=4, # use 4 threads
                      nmotifs=5, # determines how many motifs Meme will try to find\
                      meme_path = "/home/plger/bin/") # depending on your setup, you might need to specify where the meme binaries were installed

# we could also specify a background using the `neg` argument, i.e. instead of 
# looking for motifs against a random background, you can specify control (i.e. non-peak) sequences
```


## Example using rGADEM (doesn't require installing the meme suite)

```{r, eval=FALSE}
gadem <- rGADEM::GADEM(peak_seqs, genome=genome, nmotifs=5)
```


